writing/2019-10-05-Designing-testable-components.md

```diff
@@ -0,0 +1,471 @@
+constrain their possible effects - and thus limit what we need to test.
+
+We're interested in what is commonly referred to as *effect tracking*.
+
+Because we have gotten rid of `IO` in our refactoring above we can now choose
+which monad to evaluate our programs in. This gives us the power to limit the
+effects of the monad. We can evaluate it purely or we can evaluate it with
+only certain effects.
+
+There are many ways to do this in Haskell,
+[fused-effects](http://hackage.haskell.org/package/fused-effects),
+[extensible-effects](http://hackage.haskell.org/package/extensible-effects),
+[polysemy](http://hackage.haskell.org/package/polysemy),
+free monads, tagless final or the classic
+[MTL](http://hackage.haskell.org/package/mtl) approach.
+
+These libraries all exist for a single reasons - monads don't compose. You
+can't take two different monads (e.g. `Maybe` and `IO`) and compose them.
+The original solution to this is monad transformers (as found in the MTL lib).
+
+When choosing something - consider choosing the easiest thing you can possibly
+get away with. A friend once told me: "Whenever you write something, you should
+aim to write it in a non-clever way. Because, when you have to debug it - you're
+going to need to be twice as smart."
+
+We're going to go with a solution that has a slightly higher degree of
+boilerplate, which we're willing to do since it is more straight-forward.
+
+The first order of business at this point is to introduce you to a monad called
+"Reader". If you already know about it - you can skip ahead to [Actually
+getting rid of the manual wiring](#actually-getting-rid-of-the-manual-wiring).
+
+## Introducing reader
+We want to get rid of the manual wiring. This is were Reader comes into play.
+The easiest way to describe reader is to say that it is a monad that is able to
+read a value from its context.
+
+What's an example of something that can read a value from its context? Well, a
+function!
+
+```haskell
+getPersistUser :: Application m -> (UserName -> Password -> m UserId)
+getPersistUser app = app & persistence & persistUser
+```
+
+We could re-write this in a way where we don't explicitly have to pass the
+`Application` argument:
+
+```haskell
+getPersistUser :: MonadReader Application m => m (UserName -> Password -> m UserId)
+getPersistUser = do
+  app <- ask
+  pure $ app & persistence & persistUser
+```
+
+Unfortunately, the type signature has changed - but that is a very small price
+to pay since we can just unwrap it by doing something like:
+
+```haskell
+runReader getPersistUser app
+```
+
+(In fact this will make the `m` be `Identity` and then unwrap it for us!)
+
+We could of course run this in any monad we wanted to - like `Either a` or
+`Maybe` or `IO`. It might seem like a contrived example, but bear in mind that
+if we let all the functions that require this parameter be readers - we can
+compose them before actually running it and thus only pass the parameter once.
```

Comment:
I think that the composition is the main reason why we would want to use the reader pattern. It would be nice to have an example of the composition of two "services"